### PR TITLE
Consuming cpprestsdk

### PR DESCRIPTION
**Pull request by ashpatil-msft**
- Created a static library project for cpprestsdk source code and included necessary header and cpp files, preprocessor defs, etc. 
- Added required references in the other projects that will look at cpprestsdk project.
- Manually tested consuming cpprestsdk httpclient, json and others in the project and the requests work as expected using the library.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/790)

### CODE REVIEW

File: src/cpprestsdk/cpprestsdk.vcxproj.filters [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+      <Filter>src\streams</Filter>
+    </ClCompile>
+    <ClCompile Include="cpprestsdk\Release\src\http\client\http_client_winhttp.cpp">
+      <Filter>src\http\client</Filter>
+    </ClCompile>
+    <ClCompile Include="cpprestsdk\Release\src\pplx\threadpool.cpp">
+      <Filter>src\pplx</Filter>
+    </ClCompile>
+  </ItemGroup>
+</Project>
```
Comment (jsoref) [inline src/cpprestsdk/cpprestsdk.vcxproj.filters:282]: Can you include a newline at the end of the file?

File: .gitignore [hunk 2]
```diff
@@ -24,6 +24,9 @@ bld/
 [Oo]bj/
 [Ll]og/
 
+# Include cpprestsdk files in Release folder
+!src/cpprestsdk/cpprestsdk/Release
```
Comment (yao-msft) [inline .gitignore:28]: Remove, This will prevent us from updating cpprestsdk in the future

Comment (ashpatil-msft) [reply .gitignore]: removed

File: src/AppInstallerCLI.sln [hunk 3]
```diff
@@ -428,6 +430,30 @@ Global
...
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Debug|x86.ActiveCfg = Debug|Win32
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Debug|x86.Build.0 = Debug|Win32
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Fuzzing|ARM.ActiveCfg = Debug|Win32
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Fuzzing|ARM.Build.0 = Debug|Win32
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Fuzzing|ARM64.ActiveCfg = Debug|Win32
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Fuzzing|ARM64.Build.0 = Debug|Win32
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Fuzzing|x64.ActiveCfg = Debug|x64
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Fuzzing|x64.Build.0 = Debug|x64
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Fuzzing|x86.ActiveCfg = Debug|Win32
+		{866C3F06-636F-4BE8-BC24-5F86ECC606A1}.Fuzzing|x86.Build.0 = Debug|Win32
```
Comment (yao-msft) [inline src/AppInstallerCLI.sln:448]: I don't think cpprestsdk will be used in fuzzing tests, fuzzing tests are for yaml fuzzing. can you double check?

File: src/AppInstallerCLI/AppInstallerCLI.vcxproj [hunk 4]
```diff
@@ -125,10 +125,10 @@
   <ItemDefinitionGroup Condition="'$(Configuration)'=='Debug'">
     <ClCompile>
       <Optimization>Disabled</Optimization>
-      <PreprocessorDefinitions>_DEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
-      <AdditionalIncludeDirectories Condition="'$(Configuration)|$(Platform)'=='Debug|ARM'">$(ProjectDir);..\AppInstallerCLICore\Public\;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
-      <AdditionalIncludeDirectories Condition="'$(Configuration)|$(Platform)'=='Debug|ARM64'">$(ProjectDir);..\AppInstallerCLICore\Public\;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
-      <AdditionalIncludeDirectories Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">$(ProjectDir);..\AppInstallerCLICore\Public\;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <PreprocessorDefinitions>_NO_ASYNCRTIMP;_DEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
Comment (yao-msft) [inline src/AppInstallerCLI/AppInstallerCLI.vcxproj:128]: >_NO_ASYNCRTIMP; [](start = 31, length = 15)

Should this be in the cpprestsdk.vcxproj ?

File: src/AppInstallerCLI/AppInstallerCLI.vcxproj [hunk 5]
```diff
@@ -194,7 +195,7 @@
       <EnableCOMDATFolding>true</EnableCOMDATFolding>
       <OptimizeReferences>true</OptimizeReferences>
       <GenerateWindowsMetadata>false</GenerateWindowsMetadata>
-      <AdditionalDependencies Condition="'$(Configuration)'=='Release'">wininet.lib;shell32.lib;winsqlite3.lib;shlwapi.lib;icuuc.lib;icuin.lib;urlmon.lib;Advapi32.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalDependencies Condition="'$(Configuration)'=='Release'">wininet.lib;shell32.lib;winsqlite3.lib;shlwapi.lib;icuuc.lib;icuin.lib;urlmon.lib;Advapi32.lib;bcrypt.lib;winhttp.lib;$(SolutionDir)$(PlatformShortname)\$(Configuration)\cpprestsdk\cpprestsdk.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Comment (yao-msft) [inline src/AppInstallerCLI/AppInstallerCLI.vcxproj:198]: >$(SolutionDir)$(PlatformShortname)\$(Configuration)\cpprestsdk\cpprestsdk.lib; [](start = 190, length = 78)

I think it's automatically included by adding as a project reference? Like jsoncpp, libyaml, etc

**Comment by yao-msft** [comment]
Let's modify the AppInstallerCLI.vcxproj, AppInstallerCLITests.vcxproj and other vcxprojs when you start to actually using the cpprestsdk. Since I think most of the changes are unnecessary. For example, I don't think AppInstallerCLi.vcxproj needs to have those additional includes.

**Comment by yao-msft** [comment]
> <?xml version="1.0" encoding="utf-8"?>

wingetutil.vcxproj probably does not need to be changed, I don't see a usage for cpprestsdk in wingetutil.

---
Refers to: src/WinGetUtil/WinGetUtil.vcxproj:1 in 9eb7179. [](commit_id = 9eb71799c4eef9ec48982ed221f01300729c9826, deletion_comment = False)

File: src/AppInstallerCLI/AppInstallerCLI.vcxproj (hunk 4 above)
Comment (ashpatil-msft) [reply src/AppInstallerCLI/AppInstallerCLI.vcxproj:128]: It needs to be in the consuming project as well.

Comment (ashpatil-msft) [reply src/AppInstallerCLI/AppInstallerCLI.vcxproj:128]: Actually, no. We need it in the AppIntallerRepositoryCore only

---
In reply to: [592933244](https://github.com/microsoft/winget-cli/pull/790#discussion_r592933244) [](ancestors = 592933244)

File: src/AppInstallerCLI/AppInstallerCLI.vcxproj (hunk 5 above)
Comment (ashpatil-msft) [reply src/AppInstallerCLI/AppInstallerCLI.vcxproj]: removed

**Review by JohnMcPMS — APPROVED** [review]
:shipit:

**Review by yao-msft — APPROVED** [review]
:shipit:

**Pull request merged**